ci: Enable UBSan for 'longlong' builds in CI, add stack size for sanitizer builds.

projectgus opened the pull request:
### Summary

* Spun out from #17734 - enable UBSan in CI as suggested by @jepler 
* Scale up the unix port stack sizes substantially when sanitizers are enabled. The size of some MicroPython C stack frames with UBSan and GCC 12 (as used in CI) appears to be quite high. Whatever causes this seems to be fixed in later GCC.
* To support uniform scaling of stack sizes, added some macros to apply the same stack size multiplier to both the main thread and the default thread stack size on unix port. Previously the unix port main thread stack was doubled on ARM builds but the default thread stack size was not, now they are both doubled (and the same macros are used when sanitizers are enabled).
* Added a custom MicroPython macro to mark if sanitizers are enabled. As far as I can tell there's no way to programmatically detect if sanitizers are enabled on gcc<14, the other macros are all Clang-only.
* Rewrote the scaling mechanism for the `stress/recurse_iternext.py` test, as this test fails on sanitizer builds with the increased stack. (Sanitizer build stack usage is only increased due to certain C function calls, but this test - I think - consumes stack in a way which isn't impacted by sanitizers so it can now handle many more iterations compared to a standard build). This took a while to find a good approach, but what's there seems more accurate than the previous heuristic and doesn't take noticeably more time to execute.

### Testing

* Tested in CI and by running CI functions locally with an ubuntu-22.04 container and with my local newer GCC.
* Re-ran the updated `stress/recurse_iternext.py` test on unix port with and without sanitizers, esp32 port with SPIRAM, and esp8266 port.

### Trade-offs and Alternatives

* Could leave UBSan disabled in CI for the 'longlong' variant, but (as demonstrated) it is useful to find bugs.

*This work was funded through GitHub Sponsors*

projectgus committed:
tools/ci: Add UBSan to longlong CI build.

Also rewrite the sanitizer argument variables to not assume a variant.

longlong variant currently fails in this config, due to a bug fixed
in follow-up commit.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

dpgeorge commented:
> This PR with the cherry-picked commit also hit "RuntimeError: maximum recursion depth exceeded" when testing longlong+UBSan

So maybe there just needs to be more stack overall, rather than a change in stack margin?  Maybe `ports/unix/mpthreadport.c:mp_thread_create` just needs to have a larger minimum stack size for sanitizer builds?

projectgus committed:
py/misc: Add a way to detect sanitizer builds.

Clang and gcc>=14 can use __has_feature() to detect if a sanitizer
is enabled, but older GCC has no mechanism - need to set a macro
explicitly for this to be recognised.

Necessary for increasing some resource limits in sanitizer builds.
Important not to use to avoid real issues!

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus committed:
unix: Increase stack sizes if running with sanitizers.

The specific problem seems to be that the runtime "Python stack frame"
function call is several times more expensive in stack usage when running
with UBSan on older GCC (observed on gcc 11.4 as used in CI, would get
'RuntimeError: maximum recursion depth exceeded' when running some tests
with UBSan enabled.)

Other stack usage (i.e. from pushing things on the stack in Python) stays
the same. Whatever causes the usage seems to be mostly gone in later GCC
versions.

Includes a refactor to apply the same stack size multipliers
for the default thread stack size same as the main stack size.

This goes in a new port-specific header as it depends on macros
in misc.h, so can't be in mpconfigport.h.

A side effect of this is that the default thread stack size is
now doubled on ARM, same as the main stack size.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus committed:
tests/stress/recursive_iternext: Rewrite to find its own limit.

Necessary on the unix port when running with sanitizers, as the newly
increased stack size can run all tests at N=5000 without raising
RuntimeError, and increasing N to fix this causes issues on other
configurations.

This way the test progressively builds a deeper data structure until it
fails with RuntimeError. This is theoretically slower, but not noticeably
so in reality.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus commented:
@jepler Curious of your opinion on this PR, if you have any time to look?

dpgeorge commented:
I tested the new `recursive_iternext.py` test on PYBD_SF6, where it previously failed when used with the native emitter, and it now passes.

Also tested on PYBD_SF2, PYBV10, RPI_PICO.  It passes on all of those.  Very good!

jepler reviewed (commented):
I'm in favor of it.

I looked into alternatives for detecting ubsan/asan and while gcc documentation shows they were grumpy at having to implement `__has_feature`, it does work as needed.

dpgeorge commented on ports/unix/stack_size.h:
```diff
@@ -0,0 +1,29 @@
+#ifndef MICROPY_INCLUDED_UNIX_STACK_SIZE_H
```
Please add a copyright and license header.

projectgus replied:
Oops, rebased and added!

dpgeorge commented:
I retested the revised `recursive_iternext.py` test on PYBD_SF2 and PYBD_SF6, in bytecode, mpy and native mode, and it passes.

Apart from the above comment about the license, this looks good to go in.

dpgeorge reviewed (approved):
Thanks for updating.
